heron/scheduler-core/src/java/com/twitter/heron/scheduler/LaunchRunner.java

```diff
@@ -134,40 +131,36 @@ public Boolean call() {
     // Set topology def first since we determine whether a topology is running
     // by checking the existence of topology def
     // store the trimmed topology definition into the state manager
+    // TODO(rli): log-and-false anti-pattern is too nested on this path. will not refactor
     result = statemgr.setTopology(trimTopology(topology), topologyName);
     if (result == null || !result) {
-      LOG.severe("Failed to set topology definition");
-      return false;
+      throw new LauncherException("Failed to set topology definition", topologyName);
     }
 
     result = statemgr.setPackingPlan(createPackingPlan(packedPlan), topologyName);
     if (result == null || !result) {
-      LOG.severe("Failed to set packing plan");
       statemgr.deleteTopology(topologyName);
-      return false;
+      throw new LauncherException("Failed to set packing plan", topologyName);
     }
 
     // store the execution state into the state manager
     ExecutionEnvironment.ExecutionState executionState = createExecutionState();
 
     result = statemgr.setExecutionState(executionState, topologyName);
     if (result == null || !result) {
-      LOG.severe("Failed to set execution state");
       statemgr.deletePackingPlan(topologyName);
       statemgr.deleteTopology(topologyName);
-      return false;
+      throw new LauncherException("Failed to set execution state", topologyName);
     }
 
     // launch the topology, clear the state if it fails
     if (!launcher.launch(packedPlan)) {
       statemgr.deleteExecutionState(topologyName);
       statemgr.deletePackingPlan(topologyName);
       statemgr.deleteTopology(topologyName);
-      LOG.log(Level.SEVERE, "Failed to launch topology");
-      return false;
+      throw new LauncherException("Failed to launch topology", topologyName);
     }
 
     return true;
```

Comment:
that means we need to change to `Callable<Void>` [here](https://github.com/twitter/heron/pull/1571/files/39d58c0e1423bd86d4cd29f968f2804412f75e3f#diff-bc651b82b5edf290148aedd84d4b8633R37). And we have to return a `null` at the very end.